### PR TITLE
Minimal implementation of `download_dataset_parallel` 

### LINKED ISSUE #27: Remove 4 corners logic
Currently we check the 4 corners value to decide on the image_id. We should directly
- take the one given by the metadata
- using this 4 corners logic as a sanity check raising an Exception when it is not the one in the majority of the corners

### LINKED ISSUE #46: Support downsampling in `get_image`
We could use downsampled images for the synchronization and introduce a speedup.

https://help.brain-map.org/display/api/Downloading+an+Image#DownloadinganImage-ImageDownloadService

### PR DESCRIPTION
**Pull request by jankrepl**
Closes #27 and closes #46

# What was done
- [x] Created 20 `json` files containing the "ground truth" mapping between p, i, r and x,y
- [x] Implemented `get_transform_parallel` that creates the final `DisplacementField` given the 2D and 3D matrices
- [x] Implemented `download_dataset_parallel` that download an entire dataset using the `get_transform_parallel`
- [x] Wrote tests that are runnable without internet 
- [x] Image downsampling. If one provides `ds_r > 1` (downsampling in the refernence space) there is no need to actually download the entire image. A parameter `ds_i` (downsampling of the image) is introduced that does this (save a lot of time). We should verify the quality for different `ds_i`.

# What was not done
1. If we accept this PR, we can remove a big portion of the code inside of `atldld/sync.py` since it is not used 
2. The 3D transform also provides information on how the 3rd axis (that is fixed) was mapped. However, currently this information is thrown away. In theory it can be used to understand the tilt the dataset has.

# I wanna know more!
### Why do you use the term "parallel"
I just wanted to stress that we make an assumption that the entire slice is parallel with respect to 1 of the 3 axes (coronal, sagittal, transverse). In reality, this is not the case and the slices are not perfectly parallel. Anyway, if you have an idea for a better naming feel free to share it. Note that in the future we might implement more general download logic that does not make this assumption.

### So what is the algorithm/pipeline?
I tried to explain it in the docstring of `dowload_dataset_parallel`.

### How is it better than what we currently have
* Implemented in a couple of lines of code
* Works out of the box for all 3 axes (coronal, sagittal, transverse)
* Fast
* No crazy section number checking logic (4 corners)
* Testable without internet

### How did you create the `json`'s?
See below the exact script. TLDR, I took 5 coronals dataset and 5 sagittal datasts and randomly sampled 2 points in them. And then I simply queried the corresponding `p, i, r` (+ metadata)

<details>
  <summary>Click to expand the script!</summary>

```python
import json
import pathlib
import sys

import numpy as np
import requests

from atldld.utils import (
    pir_to_xy_API_single,
    get_2d_bulk,
    get_3d,
)

CONFIG = {
    "coronal": [
        717,
        923,
        1357,
        71717640,
        77371835,
    ],
    "sagittal": [
        75457491,
        75457580,
        75492803,
        79913385,
        100055064,
    ],
}


refspace_size = (13200, 8000, 11400)

output_folder = pathlib.Path.cwd() / "tests" / "data" / "sync" / "pir_to_xy"
np.random.seed(10)

ds_f = 50
n_points_per_dataset = 2

for axis, datasets in CONFIG.items():
    for dataset_id in datasets:
        for _ in range(n_points_per_dataset):
            p_ = np.random.randint(0, int(refspace_size[0] / ds_f))
            i_ = np.random.randint(0, int(refspace_size[1] / ds_f))
            r_ = np.random.randint(0, int(refspace_size[2] / ds_f))
            p, i, r = ds_f * p_, ds_f * i_, ds_f * r_

            if axis == "coronal":
                assert i % ds_f == 0
                assert r % ds_f == 0
                reference_space = 9
            else:
                assert p % ds_f == 0
                assert i % ds_f == 0
                reference_space = 10

            x, y, section_number, section_img_id = pir_to_xy_API_single(
                p,
                i,
                r,
                dataset_id,
                reference_space=reference_space,
            )

            matrix_3d = get_3d(
                dataset_id,
                ref2inp=True,
                add_last=False,
                return_meta=False,
            )

            metadata_2d = get_2d_bulk(
                dataset_id,
                ref2inp=True,
                add_last=False,
            )
            matrix_2d = metadata_2d[section_img_id][0]

            print(dataset_id, x, y, section_img_id, matrix_2d, matrix_3d)

            res = {
                "dataset_id": dataset_id,
                "image_id": section_img_id,
                "section_number": section_number,
                "axis": axis,
                "p": p,
                "i": i,
                "r": r,
                "x": x,
                "y": y,
                "affine_2d": matrix_2d.tolist(),
                "affine_3d": matrix_3d.tolist(),
            }

            path = output_folder / f"{dataset_id}_{p}_{i}_{r}.json"

            with path.open("w") as f:
                json.dump(res, f, indent=2)

```

</details>

### How do I try it?
It has exactly the same API as before

```python
import matplotlib.pyplot as plt
from atldld.sync import download_dataset_parallel


gen = download_dataset_parallel(
    dataset_id=75457491,
    downsample_ref=25,
    downsample_img=4,
)

for img_id, section_coordinate, img, df in gen:
    print(img_id, section_coordinate)
    continue
    # Optional
    _, axs = plt.subplots(2)
    axs[0].imshow(img)
    axs[1].imshow(df.warp(img))
    plt.show()
```
The above script takes ~85 seconds (~4.5 seconds per image - there are 20 of them) with an empty cache (everything is dowloaded).

### CODE REVIEW

**Comment by jankrepl** [comment]
@EmilieDel @Stannislav Feel free to review it!!!

File: src/atldld/sync.py [hunk 1]
```diff
@@ -519,6 +520,174 @@ def corners_sagittal(r, dataset_id):
...
+    include_expression: bool = False,
+    ds_i: int = 0,
+) -> Generator[
+    Union[
+        Tuple[int, float, np.ndarray, DisplacementField],
+        Tuple[int, float, np.ndarray, DisplacementField, np.ndarray],
+    ],
+    None,
+    None,
+]:
```
Comment (Stannislav) [inline src/atldld/sync.py:599]: - I think it's recommended to just annotate `float` instead of `Union[int, float]` because [`float` already implies `int`](https://www.python.org/dev/peps/pep-0484/#the-numeric-tower)
- That's a huge type annotation 😅 I guess there's nothing we can do about it at this point. Maybe some refactoring can be done in the future to avoid the ambiguous return type.

By the way for the cases like this where parameter/return types depend on each other there's the `overload` decorator in `typing`. It's probably an overkill to add overloads here. But just for info, it would look something like this:
```python
@overload
def download_dataset_parallel(
    dataset_id: int,
    ds_r: int = ...,
    detection_xy: Tuple[float, float] = ...,
    include_expression: Literal[False] = ...,
    ds_i: int = ...,
) -> Generator[
    Tuple[int, float, np.ndarray, DisplacementField],
    None,
    None,
]:
    ...


@overload
def download_dataset_parallel(
    dataset_id: int,
    ds_r: int = ...,
    detection_xy: Tuple[float, float] = ...,
    include_expression: Literal[True] = ...,
    ds_i: int = ...,
) -> Generator[
    Tuple[int, float, np.ndarray, DisplacementField, np.ndarray],
    None,
    None,
]:
    ...
```

File: src/atldld/sync.py [hunk 2]
```diff
@@ -519,6 +520,174 @@ def corners_sagittal(r, dataset_id):
...
+        dataset_id,
+        ref2inp=True,
+        add_last=False,
+    )
+    matrix_3d = get_3d(
+        dataset_id,
+        ref2inp=True,
+        add_last=False,
+        return_meta=False,
+    )
```
Comment (Stannislav) [inline src/atldld/sync.py:659]: My PyCharm is suggesting that one should remove `add_last=False` and `return_meta=False` because the values are equal the default values.

File: src/atldld/utils.py [hunk 3]
```diff
@@ -111,6 +115,7 @@ def get_image(image_id, folder=None, expression=False):
 
     # Create full path
     additional_specifier = "_expression" if expression else ""
+    additional_specifier += f"_{downsample}"
```
Comment (Stannislav) [inline src/atldld/utils.py:118]: Maybe the line below (where the path is constructed as a bare string) is the reason for the bug under point 2 in #41? I feel like we should use `pathlib.Path` here (or `os.path`, but personally I prefer `pathlib`)

So, personally, I would write something like this:
```python
file_name = f"{image_id}-{downsample}"
if expression:
    file_name += "-expression"
path = pathlib.Path(folder) / (file_name + ".jpg")
```

File: src/atldld/utils.py [hunk 4]
```diff
@@ -126,8 +131,15 @@ def get_image(image_id, folder=None, expression=False):
...
+        options = []
+
         if expression:
-            image_url += "?view=expression"
+            options.append("view=expression")
+
+        if downsample:
+            options.append(f"downsample={downsample}")
+
+        image_url += "?" + "$".join(options)
```
Comment (Stannislav) [inline src/atldld/utils.py:142]: Is it not `"&"` instead of `"$"` in `join`?

File: tests/conftest.py [hunk 5]
```diff
@@ -16,9 +16,16 @@
...
 
+import json
+import pathlib
+
 import numpy as np
 import pytest
 
+DATA_FOLDER = pathlib.Path(__file__).parent / "data"
+PIR_TO_XY_FOLDER = DATA_FOLDER / "sync" / "pir_to_xy"
+PIR_TO_XY_RESPONSES = sorted(PIR_TO_XY_FOLDER.rglob("*json"))
```
Comment (Stannislav) [inline tests/conftest.py:27]: The `PIR_TO_XY_FOLDER` has only the response JSON files in it, right? Because reading `rglob("*json")` makes one think there might be subfolders and non-JSON files. Maybe `sorted(PIR_TO_XY_FOLDER.iterdir())` would also do the trick?

**Review by Stannislav — COMMENTED** [review]
Thanks @jankrepl  :)

File: src/atldld/sync.py [hunk 6]
```diff
@@ -519,6 +520,176 @@ def corners_sagittal(r, dataset_id):
...
+    Parameters
+    ----------
+    slice_coordainte : float
+        Value of the `axis` coordinate at which the image was sliced.
+    matrix_2d : np.array
+        Matrix of shape `(3, 3)` representing a 2D linear transformation.
+        Note that the last row contains the `homogeneous` coordinates.
+    matrix_3d : np.array
+        Matrix of shape `(4, 4)` represnting a 3D linear transformation.
+        Note that the last row contains the `homogeneous` coordinates.
```
[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py]: Good points. Btw I think you reviewed an outdated version but whatever:)

File: src/atldld/sync.py [hunk 7]
```diff
@@ -519,6 +520,174 @@ def corners_sagittal(r, dataset_id):
...
+
+    coords_temp = np.ones((3, n_pixels))
+    coords_temp[[0, 1]] = (matrix_3d @ coords_ref)[:2]  # (3, 4) x (4, n_pixels)
+
+    coords_img = matrix_2d @ coords_temp  # (2, 3) x (3, n_pixels)
+
+    tx = coords_img[0].reshape(grid_shape) / (2 ** ds_i)
+    ty = coords_img[1].reshape(grid_shape) / (2 ** ds_i)
+
+    df = DisplacementField.from_transform(tx, ty)  # `from_transform` not annotated
```
[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py:581]: Ideally it should imply it automatically, right?

File: src/atldld/sync.py [hunk 8]
```diff
@@ -519,6 +520,176 @@ def corners_sagittal(r, dataset_id):
...
+    DisplacementField
+        Displacement field representing the transformation between the
+        reference space and the image. Note that one can directly use it
+        to register raw histological images to the reference space.
+    """
+    refspace = (  # order matters
+        ("coronal", 13200),
+        ("transverse", 8000),
+        ("sagittal", 11400),
+    )
```
[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py:562]: Sure! Great idea:) However, I would suggest doing it outside of this PR.

File: src/atldld/sync.py [hunk 9]
```diff
@@ -519,6 +520,174 @@ def corners_sagittal(r, dataset_id):
...
+    axis_fixed = [i for i, a in enumerate(refspace) if a[0] == axis][0]
+    axes_variable = [i for i, a in enumerate(refspace) if a[0] != axis]
+
+    grid_shape = [refspace[i][1] // ds_r for i in axes_variable]
+    n_pixels = np.prod(grid_shape)
+
+    coords_ref = np.ones((4, n_pixels))
```
[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py:569]: I guess in theory one could do it without flattening. However, IMO flattening across the pixels makes the code more readable and the matrix multiplications are easier to read.

However, I would be willing to change it in case there is a major performance benefit

File: src/atldld/utils.py (hunk 4 above)
Comment (jankrepl) [reply src/atldld/utils.py:142]: Great catch! Thank you!

File: tests/conftest.py (hunk 5 above)
Comment (jankrepl) [reply tests/conftest.py:27]: Great point!

File: src/atldld/utils.py (hunk 3 above)
Comment (jankrepl) [reply src/atldld/utils.py:118]: Yeh, I might as well do the fix in this PR. It is tiny:) Thanks.

File: src/atldld/sync.py [hunk 10]
```diff
@@ -519,6 +520,174 @@ def corners_sagittal(r, dataset_id):
...
+    coords_ref[axis_fixed] *= slice_coordinate
+    coords_ref[axes_variable] = np.indices(grid_shape).reshape(2, -1) * ds_r
+
+    coords_temp = np.ones((3, n_pixels))
+    coords_temp[[0, 1]] = (matrix_3d @ coords_ref)[:2]  # (3, 4) x (4, n_pixels)
+
+    coords_img = matrix_2d @ coords_temp  # (2, 3) x (3, n_pixels)
+
+    tx = coords_img[0].reshape(grid_shape) / (2 ** ds_i)
+    ty = coords_img[1].reshape(grid_shape) / (2 ** ds_i)
```
[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py:579]: Yeh, totally agree that there is a lot of magic happening. I was thinking of writing a dedicated page in the docs. However, first I wanted to get your initial feedback.

TLDR
1. We generate coordinate grid in the reference space (parallel to one of the axis)
2. We downsample it by `ds_r`
3. For each point in the downsampled reference grid (p, i, r) we find the corresponding (x, y) in the image. We do this by  the 2 matrix multiplications.
4. The user might have downloaded a downsampled image, if that is the case we need scale our `(tx, ty)` accordingly. For example, if we downloaded an image with `ds_i=2`  whatever was the coordinate (16, 64) in the raw (undownsampled) image will become (16 / 2 ** 2, 64 / 2 **2) = (4, 8) in the new image. The mapping we get is always with respect to the raw image. Note that this step actually allows for huge speedups when it comes to downloading.

File: src/atldld/utils.py (hunk 4 above)
Comment (jankrepl) [reply src/atldld/utils.py]: Fixed in f53237d853eee51788a313665e01a6a10373a203

File: tests/conftest.py (hunk 5 above)
Comment (jankrepl) [reply tests/conftest.py]: Fixed in 4e978e1a6279dd41940993f361ab908a083d1796

File: src/atldld/sync.py (hunk 2 above)
Comment (jankrepl) [reply src/atldld/sync.py]: Hmmm...I guess this is not something `flake8` would catch, right? Anyway, fixed in 47fdecd6a186c272be635cb55854e139d53db7ed

File: src/atldld/sync.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py]: Fixed in a23aea4e27ad08891808bd59d5a675fba808fccf

File: src/atldld/sync.py [hunk 11]
```diff
@@ -519,6 +520,174 @@ def corners_sagittal(r, dataset_id):
...
+    matrix_3d: np.ndarray,
+    axis: str = "coronal",
+    ds_r: int = 1,
+    ds_i: int = 0,
+) -> DisplacementField:
+    """Compute displacement field between the reference space and the image.
+
+    Parameters
+    ----------
+    slice_coordainte : float
```
[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py]: Fixed in `a23aea4e27ad08891808bd59d5a675fba808fccf`

File: src/atldld/sync.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (EmilieDel) [reply src/atldld/sync.py:579]: Same as @Stannislav. Thanks for the comment, @jankrepl, it helps!

Do you think some comments in the code could be helpful to keep track of the different steps ?

File: src/atldld/sync.py (hunk 1 above)
Comment (jankrepl) [reply src/atldld/sync.py]: Interesting! Thank you! Something to keep in mind! Anyway, I fixed the `float` thing here 574521fc8d64d604e7469e00c534d48db5352f46

File: tests/data/sync/pir_to_xy/100055064_12500_4800_10800.json [hunk 12]
```diff
@@ -0,0 +1,43 @@
+{
```
Comment (EmilieDel) [inline tests/data/sync/pir_to_xy/100055064_12500_4800_10800.json:1]: Just a small suggestion, should we put somewhere the convention of the naming of those files ? `dataset_id_p_i_r` ?

**Review by EmilieDel — APPROVED** [review]
Really nice PR, thanks @jankrepl !

File: tests/data/sync/pir_to_xy/100055064_12500_4800_10800.json (hunk 12 above)
Comment (jankrepl) [reply tests/data/sync/pir_to_xy/100055064_12500_4800_10800.json:1]: I included the script that generates these files in the description. IMO it is a good enough documentation. The actual name of the file is never used. It is the contents that matter.

Comment (EmilieDel) [reply tests/data/sync/pir_to_xy/100055064_12500_4800_10800.json:1]: Sounds good for me!

File: src/atldld/sync.py [hunk 13]
```diff
@@ -519,6 +520,172 @@ def corners_sagittal(r, dataset_id):
...
+    all_image_ids = sorted(
+        metadata_2d.keys(),
+        key=lambda x: -int(metadata_2d[x][1]),
+    )
+
+    for image_id in all_image_ids:
+        p, i, r = xy_to_pir_API_single(*detection_xy, image_id=image_id)
+
+        slice_ref_coordinate = p if axis == "coronal" else r
+        affine_2d = [v[0] for k, v in metadata_2d.items() if k == image_id][0]
```
Comment (Stannislav) [inline src/atldld/sync.py:669]: Make this more readable / transparent.

File: src/atldld/sync.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py]: The corresponding issue is here #33.

File: src/atldld/utils.py (hunk 3 above)
Comment (jankrepl) [reply src/atldld/utils.py:118]: Sorry, I changed my mind:D Let's implement this in a separate PR.

File: src/atldld/sync.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Stannislav) [reply src/atldld/sync.py:581]: Well in the line below it says
```python
return df  # type: ignore
```

I think by adding the annotation I mentioned we could drop this ignore.

It seems that `mypy` cannot figure out that `DisplacementField.from_transform(tx, ty)` returns an instance of `DisplacementField` for some reason...

[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py:581]: Ultimately, the correct way to do this is to annotate all the methods inside of `DisplacementField`.  However, it is beyond the scope of this PR. So  I just wrote `# `from_transform` not annotated` and ignored the type.

Hope it makes sense.

File: src/atldld/sync.py (hunk 13 above)
Comment (jankrepl) [reply src/atldld/sync.py]: Fixed in 88dfe0115d00f212e1e225e5af52505028f3c9d7. I hope it is more readable than before:)

File: src/atldld/sync.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py]: However, you are right that it is nicer to manually provide type rather than ignore it.

Did it here 4c4ab00ca48681c9e0b8172e6d2f51a3d3aa0439

File: src/atldld/sync.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py:562]: I will close this but definitely a great idea:)

File: src/atldld/sync.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jankrepl) [reply src/atldld/sync.py:569]: I will resolve, feel free to add comments/unresolve

**Comment by jankrepl** [comment]
@Stannislav @EmilieDel Feel free to merge. I tried to address all the issues. I will try to document everything in a separate PR (see #33)

File: src/atldld/sync.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Stannislav) [reply src/atldld/sync.py:562]: Sure, agree! Created new issue: #55

File: src/atldld/sync.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Stannislav) [reply src/atldld/sync.py]: nice, thanks!

File: src/atldld/sync.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Stannislav) [reply src/atldld/sync.py:569]: Ha, that's funny, it must be personal, but I find it easier to read when it's not flattened 🙈 Maybe you prefer thinking in terms of 2D matrices? Anyway, doesn't matter so much :)

File: src/atldld/utils.py (hunk 3 above)
Comment (Stannislav) [reply src/atldld/utils.py:118]: ha ha, fair enough. Is there already an issue for this? Maybe we can copy the code snippet above over there.

Comment (jankrepl) [reply src/atldld/utils.py:118]: Yes, feel free to put it in #41. I think using `pathlib` is a necessity.

**Review by Stannislav — APPROVED** [review]
Thanks @jankrepl
